lib/THCUNN/SparseLinear.cu

```diff
@@ -0,0 +1,284 @@
+      } else {
+        THError("index out of bound. updateOutput: %ld not between 1 and %ld",
+            offset + 1, inDim);
+      }
+    }
+  }
+#else
+  THCudaTensor *colinds, *values_sel, *values;
+  THCudaIntTensor *colinds_int, *csr_int;
+
+  THCudaTensor_resize2d(state, output, batchnum, outDim);
+  // output = bias
+  for (h=0; h<batchnum; h++) {
+    THCudaTensor_select(state, output_row, output, 0, h);
+    THCudaTensor_copy(state, output_row, bias);
+  }
+
+  // Get data ready for cusparse
+  values_sel = THCudaTensor_newSelect(state, input, 2, 1);
+  values = THCudaTensor_newContiguous(state, values_sel);
+
+  colinds = THCudaTensor_newSelect(state, input, 2, 0);
+  colinds_int = THCudaIntTensor_newWithSize1d(state, nnz*batchnum);
+  THCudaIntTensor_copyCudaFloat(state, colinds_int, colinds);
+
+  // See http://docs.nvidia.com/cuda/cusparse/index.html#compressed-sparse-row-format-csr
+  csr_int = THCudaIntTensor_newWithSize1d(state, batchnum+1);
+  init_cusparse();
+  for (h = 0; h < batchnum+1; h++) {
+    THCudaIntTensor_set1d(state, csr_int, h, 1 + nnz * h);
+  }
+
+  float one = 1;
+  cusparseMatDescr_t descr=0;
+  cusparseCreateMatDescr(&descr);
+  cusparseSetMatType(descr,CUSPARSE_MATRIX_TYPE_GENERAL);
+  cusparseSetMatIndexBase(descr,CUSPARSE_INDEX_BASE_ONE);  
+  cusparseScsrmm(cusparse_handle,
+      CUSPARSE_OPERATION_NON_TRANSPOSE,
+      batchnum, outDim, inDim, nnz*batchnum,
+      &one,
+      descr,
+      THCudaTensor_data(state, values),
+      THCudaIntTensor_data(state, csr_int),
+      THCudaIntTensor_data(state, colinds_int),
+      THCudaTensor_data(state, weight), inDim,
+      &one, THCudaTensor_data(state, output), batchnum
+  );
+  cusparseDestroyMatDescr(descr); 
+  descr = 0;
+  THCudaTensor_free(state, colinds);
+  THCudaTensor_free(state, values_sel);
+  THCudaTensor_free(state, values);
+  THCudaIntTensor_free(state, colinds_int);
+  THCudaIntTensor_free(state, csr_int);
+#endif
+  THCudaTensor_free(state, output_row);
+}
+
+void THNN_CudaSparseLinear_zeroGradParameters(
+          THCState *state,
+          THCudaTensor *gradWeight,
+          THCudaTensor *gradBias,
+          THCudaTensor *lastInput)
+{
+  long h, i;
+  long outDim = THCudaTensor_size(state, gradWeight, 0);
+  long inDim = THCudaTensor_size(state, gradWeight, 1);
+  THCudaTensor *tmp = THCudaTensor_new(state);
+
+  THArgCheck(checkSize1D(gradBias, outDim), 3, "gradBias size wrong");
+  THArgCheck(checkInput(lastInput), 4,
+      "input size must be batchsize x nnz x 2");
+  THArgCheck(THCudaTensor_isContiguous(state, gradWeight), 2,
+      "gradWeight must be contiguous to use cuda");
+
+  if (!THCudaTensor_isContiguous(state, gradWeight))
+    gradWeight = THCudaTensor_newContiguous(state, gradWeight);
+  THCudaTensor_zero(state, gradBias);
+
+  long batchnum = THCudaTensor_size(state, lastInput, 0);
+  long nnz = THCudaTensor_size(state, lastInput, 1);
+
+  for (h = 0; h < batchnum; h++) {
```

Comment:
A cuda module really shouldn't have a double for loop like this. move this whole operation into a CUDA kernel as well. Ask @wickedfoo if any of his generic apply kernels can be leveraged so that you dont need to write this from scratch.